vars/buildDockerAndPublishImage.groovy

```diff
@@ -13,96 +13,73 @@ def call(String imageName, Map config=[:]) {
-                nextVersion = nextVersion + metadata
-              }
-            }
-          }
+            nextVersion = sh(script:'jx-release-version', returnStdout: true).trim()
+            if (dockerConfig.metadataFromSh != '') {
+              metadata = sh(script: "${dockerConfig.metadataFromSh}", returnStdout: true).trim()
+              nextVersion = nextVersion + metadata
+            } // if
+          } // container
           echo "Next Release Version = $nextVersion"
-        } // steps
-      } // stage
+        } // stage
+      } // if
+
 
       stage('Prepare') {
-        environment {
-          DOCKER_REGISTRY = credentials("${dockerConfig.credentials}")
-        }
-        steps {
+        withCredentials([usernamePassword(credentialsId: dockerConfig.credentials, passwordVariable: 'DOCKER_REGISTRY_PSW', usernameVariable: 'DOCKER_REGISTRY_USR')]) {
+          // Logging in on the Dockerhub helps to avoid request limit from DockerHub
           sh 'echo "${DOCKER_REGISTRY_PSW}" | img login -u "${DOCKER_REGISTRY_USR}" --password-stdin'
-          script {
-            // Retrieve Makefile
-            writeFile file: 'Makefile', text: makefileContent
-          }
-        } // steps
+
+          // The makefile to use must come from the pipeline to avoid a nasty user trying to exfiltrate data from the build
+          // Even though we have mitigation trhough the multibranch job config only allowed to build PRs from repo. contributors
+          writeFile file: 'Makefile', text: makefileContent
+        } // withCredentials
       } // stage
 
       stage("Lint") {
-        steps {
-          sh 'make lint'
-        } // steps
-        post {
-          always {
-            recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
-          }
-        } // post
-      } //stage
+        int lintStatus = sh(returnStatus: true, script: 'make lint')
+        recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
+        if (lintStatus > 0) {
+          currentBuild.result = 'FAILURE'
+          error('Lint Failed')
+        }
+      } // stage
+
 
       stage("Build") {
-        steps {
-          sh 'make build'
-        } // steps
+        sh 'make build'
       } //stage
 
-      stage("Test") {
-        when {
-          expression { fileExists 'cst.yml' }
-        }
-        steps {
-          sh 'make test'
-        } // steps
-      } //stage
 
-      stage("Semantic Release") {
-        when {
-          allOf {
-            expression { env.SEMANTIC_RELEASE == 'true' }
-            branch dockerConfig.mainBranch  
+      // Test step is not mandatory as not all repositories
+      if (fileExists('cst.yml')) {
+        stage("Test") {
+          if (sh(returnStatus: true, script: 'make test') > 0) {
+            currentBuild.result = 'FAILURE'
+            error('Test Failed')
           }
```

Comment:
First of all, avoid ever setting `currentBuild.result`. Throwing `error` will automatically mark the build as a failure (assuming it is not caught and suppressed). And why are you doing all this anyway?
```suggestion
          sh 'make test'
```
also fails the build if `make` failed. KISS

Reply:
👍